### PR TITLE
Fix async timeouts

### LINKED ISSUE #1075: Cypress hangs up when using done within a before hook after cy.request
### Current behavior:
When using a `done` callback within a `then` method of a `cy.request` in a `before` hook, tests won't wait for elements to appear on page (page load/redirect) and the whole test-suite will hang-up (it never finishes).

image

When not using the `done` function in the hook, everything works as expected = the `cy.get('button')`  waits for the page to load.

### Desired behavior:
The test behavior should be independent of the usage of the `done` function inside the `before` hook.

### How to reproduce:
Try running my simple test code with a page that take a while to load.

#### Test code:

```js
describe('A test with done function chained with a request', function () {
  before(function (done) {
    cy.request('/')
      .then(function () {
        done();
      });
  });

  it('should never get into a never-ending loop', function () {
    cy.login() // Our custom login method
    cy.visit('/');

    // Cypress doesn't wait for this elements
    cy.get('button')
    // Click does throw an error, but the test won't terminate and will never timeout
      .click();
  });
});
```

- Operating System: Ubuntu 17.10
- Cypress Version: 1.2.0
- Browser Version: Google Chrome 63.0.3239.84 (Official Build) (64-bit)

### LINKED ISSUE #2478: Mixing cy.on()/cy.once() with a one-argument it() function causes test to hang
Giving the lambda passed into the `it()` block a single argument will cause Cypress to pass a function in that must be called for the test to finish.

If you use the `cy.on()` command inside an `it()` block with a single-argument function, the test will hang when it reaches `cy.on()`.

HTML:

```html
<html>
  <body>
    <div id='covered' style='position:absolute;background:red;'>asdf</div>
    <div id='uncovered' style='position:absolute;background:blue;'>asdf</div>
  </body>
</html>
```

Test:

```javascript
describe('Test', () => {
  it('Tests done()', done => {
    let shouldCallDone = false;

    cy.log('This log should happen.');

    cy.once('fail', (err) => {
      shouldCallDone = true;
    });

    cy.get('#covered').click({ timeout: 1000 });

    cy.log('This log should also happen, but it doesn\'t.');

    cy.get('html').then(() => {
      if (shouldCallDone)
        done();
      else
        throw "done() was never called."
    });
  });

  it('Tests cy.log', () => {
    cy.log('This log should happen too, but it also doesn\'t.');
  });
});

```

### Current behavior:

image

Notice how the test has been running for almost 14 minutes.

### Desired behavior:

This code should not cause Cypress to hang.

### Versions

Cypress 3.1.0
Windows 10
Electron 59

### LINKED ISSUE #3349: Test runs infinitely with combination of async error, 'fail' event, and done callback
### Current behavior:

Test will hang if there's an async error, use of the `fail` event to catch the failing test, and a `done` callback. Since the command resolves before the error is thrown, the `fail` event is not fired and the `done` never gets invoked.

This is primarily an internal issue since we use the `fail` event to test Cypress. can be seen in the [failing driver test](https://circleci.com/gh/cypress-io/cypress/57375#tests/containers/1) in [this PR](https://github.com/cypress-io/cypress/pull/3244). That test is using new and changed aspects of ~4.0~ 5.0, but it can reproduced in the current version of Cypress using the steps below.

### Desired behavior:

The `fail` event should be fired, which will allow the test to finish.

### Steps to reproduce: (app code and test code)

```html
// visit_error.html
<!DOCTYPE html>
<html>
  <script>
    setTimeout(() => {
      foo.bar()
    })
  </script>
</body>
</html>
```

```js
// spec
it('tests visit error', (done) => {
  cy.on('fail', () => {
    done()
  })

  cy.visit('/visit_error.html')
})
```

### Versions

Cypress 3.1.5
Chrome 71
MacOS 10.14.2

### LINKED ISSUE #4790: beforeEach with done causes tests to hang indefinitely when done() called in promise.then()
### Current behaviour:

When using the `beforeEach` with callback (i.e. `beforeEach(done => promise.then(() => done())`). The cypress runner hangs when a test fails.

### Desired behavior:

When a test fails, move to the next one and in the end, turn off the runner.
At very least, implement timeout that won't allow runner to be running forever.

### Steps to reproduce: (app code and test code)

Sorry I didn't know there is tiny version. Hopefully, tiny change to kitchensink should be ok
https://github.com/cypress-io/cypress-example-kitchensink/compare/master...viktomas:before-each-hangs?expand=1

This is the whole change in https://github.com/cypress-io/cypress-example-kitchensink that reproduces the issue.

```diff
diff --git a/cypress/integration/examples/actions.spec.js b/cypress/integration/examples/actions.spec.js
index 61c69fc..18c9f0f 100644
--- a/cypress/integration/examples/actions.spec.js
+++ b/cypress/integration/examples/actions.spec.js
@@ -1,8 +1,8 @@
 /// <reference types="Cypress" />

 context('Actions', () => {
-  beforeEach(() => {
-    cy.visit('http://localhost:8080/commands/actions')
+  beforeEach(done => {
+    cy.visit('http://localhost:8080/commands/actions').then(() => done())
   })

   // https://on.cypress.io/interacting-with-elements
```
### Versions

Cypress 3.4.0, osx, node 8 and node 10

## Note
Reading through https://github.com/cypress-io/cypress/issues/1075 I think these issues might be related.

### LINKED ISSUE #4805: Any argument in Mocha callback cause the test to run forever
### Current behavior:

When you use any argument in Mocha callback, it thinks the test is asynchronous and this callback means "done" to be called later.   This will cause the Test Runner to run forever.

### Desired behavior:

It would be nice to disallow it or to throw an error.

### Steps to reproduce: (app code and test code)

Causing the issue:

```js
    it.only('should return the all accounts when search is clicked with empty search input', (callback) => {
        cy
            .get(AccountsPage.searchButton)
            .click()
            .get(AccountsPage.totalAccountsFoundText)
            .invoke('text')
            .should('match', /^[0-9]+ accounts? found$/);
    });
```

vs working version:

```js
    it.only('should return the all accounts when search is clicked with empty search input', () => {
        cy
            .get(AccountsPage.searchButton)
            .click()
            .get(AccountsPage.totalAccountsFoundText)
            .invoke('text')
            .should('match', /^[0-9]+ accounts? found$/);
    });
```

### Versions
Cypress: 3.4.0
operating system: Windows8.1
browser: Chrome 75

### PR DESCRIPTION
**Pull request by kuceb**
#### fix async tests with `done` callbacks never timing out after cypress command
- fix #3349
- fix #1075
- fix #4790
- fix #2478
- fix #4805 

### user facing changes
- tests with a `done` callback will now timeout after `defaultCommandTimeout` after all commands have run

### tasks
- [x] [review] remove adding new config value: use `defaultCommandTimeout` as `asyncTestTimeout` (for now)
- [x] waiting on decaf PR #5096 
- [x] [review] add e2e test for async timeout failure (edit existing)
- [x] [review] remove runner_spec.js tests, keep test on :92
### Pre-merge Tasks

- [x] Have tests been added/updated for the changes in this PR?
- [x] Has the original issue been tagged with a release in ZenHub?

### CODE REVIEW

File: packages/driver/test/cypress/integration/cypress/runner_spec.js [hunk 1]
```diff
@@ -32,4 +41,79 @@ describe('src/cypress/runner', () => {
       expect(pending[1].title).to.eq('is pending 2')
     })
   })
+
+  context('async timeouts', () => {
```
Comment (jennifer-shehane) [inline packages/driver/test/cypress/integration/cypress/runner_spec.js]: Move all of the tests from `runner_spec.js` in this suits to an e2e test and actually evaluate the exit code. We want to try and avoid 'modifying reality' by pulling out the error then returning false if we can.

**Review by jennifer-shehane — CHANGES_REQUESTED** [review]

File: packages/driver/test/cypress/integration/cypress/runner_spec.js (hunk 1 above)
Comment (Bkucera) [reply packages/driver/test/cypress/integration/cypress/runner_spec.js]: kept these, but added e2e test. LMK if i should remove

**Review by jennifer-shehane — APPROVED** [review]
Yay this looks great! Will merge in after the keys are fixed. 👍

**Comment by MadLittleMods** [comment]
Does this PR also fix https://github.com/cypress-io/cypress/issues/4742 ?

**Comment by ryan-snyder** [comment]
@MadLittleMods It shouldn't, no. I believe that they're seperate issues

**Pull request merged**